File: taurus-protect-sdk-python/taurus_protect/credentials.py
"""Credentials - authentication mechanism for ProtectClient."""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Callable, Optional

from taurus_protect.crypto.tpv1 import TPV1Auth
from taurus_protect.errors import ConfigurationError


class Credentials(ABC):
    """The authentication mechanism for a ProtectClient: either static TPV1-HMAC
    (api key + secret) or a Bearer token (static, or resolved per request for
    rotating/per-caller tokens).

    Build one with :meth:`api_key`, :meth:`bearer_token`, or
    :meth:`bearer_token_provider` and pass it to ``ProtectClient.create``.
    SuperAdmin keys are required regardless of the mechanism.
    """

    @staticmethod
    def api_key(api_key: str, api_secret: str) -> "Credentials":
        """TPV1-HMAC credentials (a static shared-service identity).

        Args:
            api_key: The API key.
            api_secret: The API secret as a hex-encoded string.
        """
        return _ApiKeyCredentials(api_key, api_secret)

    @staticmethod
    def bearer_token(token: str) -> "Credentials":
        """A single static Bearer token (the "Authorization: Bearer" header)."""
        if not token:
            raise ConfigurationError("bearer_token cannot be empty")
        return _BearerCredentials(lambda: token)

    @staticmethod
    def bearer_token_provider(provider: Callable[[], str]) -> "Credentials":
        """A Bearer token resolved per request from ``provider()`` (for
        rotating/per-caller tokens)."""
        if provider is None:
            raise ConfigurationError("bearer_token_provider cannot be None")
        return _BearerCredentials(provider)

    @abstractmethod
    def _build_rest_client(self, config: Any) -> Any:
        """Build the authenticating OpenAPI REST client from the configuration."""

    def close(self) -> None:
        """Wipe any sensitive material held by the credentials. No-op by default."""


class _ApiKeyCredentials(Credentials):
    def __init__(self, api_key: str, api_secret: str) -> None:
        if not api_key:
            raise ConfigurationError("api_key cannot be empty")
        if not api_secret:
            raise ConfigurationError("api_secret cannot be empty")
        try:
            self._auth: Optional[TPV1Auth] = TPV1Auth(api_key, api_secret)
        except ValueError as e:
            raise ConfigurationError(str(e)) from e

    def _build_rest_client(self, config: Any) -> Any:
        from taurus_protect.crypto.authenticated_rest import AuthenticatedRESTClient

        return AuthenticatedRESTClient(config, self._auth)

    def close(self) -> None:
        if self._auth:
            self._auth.close()


class _BearerCredentials(Credentials):
    def __init__(self, provider: Callable[[], str]) -> None:
        self._provider = provider

    def _build_rest_client(self, config: Any) -> Any:
        from taurus_protect.crypto.bearer_rest import BearerRESTClient

        return BearerRESTClient(config, self._provider)

File: taurus-protect-sdk-typescript/src/credentials.ts
/**
 * Authentication mechanism for a {@link ProtectClient}: either static TPV1-HMAC
 * (api key + secret) or a Bearer token (static, or resolved per request for
 * rotating/per-caller tokens).
 *
 * Build one with {@link Credentials.apiKey}, {@link Credentials.bearerToken}, or
 * {@link Credentials.bearerTokenProvider} and pass it as `credentials` to
 * {@link ProtectClient.create}. SuperAdmin keys are required regardless of the
 * mechanism.
 */

import type { Middleware } from "./internal/openapi/runtime";
import { ConfigurationError } from "./errors";
import {
  createBearerMiddleware,
  createTPV1Middleware,
  type BearerTokenProvider,
} from "./transport";

export class Credentials {
  // Builds the OpenAPI auth middleware for the chosen mechanism.
  private constructor(private readonly build: () => Middleware) {}

  /**
   * TPV1-HMAC credentials (a static shared-service identity).
   *
   * @param apiKey - The API key.
   * @param apiSecret - The API secret in hexadecimal format.
   */
  static apiKey(apiKey: string, apiSecret: string): Credentials {
    if (!apiKey) {
      throw new ConfigurationError("apiKey is required");
    }
    if (!apiSecret) {
      throw new ConfigurationError("apiSecret is required");
    }
    if (!/^[0-9a-fA-F]+$/.test(apiSecret)) {
      throw new ConfigurationError("apiSecret must be a valid hexadecimal string");
    }
    return new Credentials(() => createTPV1Middleware(apiKey, apiSecret));
  }

  /**
   * A single static Bearer token (the "Authorization: Bearer" header).
   */
  static bearerToken(token: string): Credentials {
    if (!token) {
      throw new ConfigurationError("bearerToken is required");
    }
    return new Credentials(() => createBearerMiddleware(() => token));
  }

  /**
   * A Bearer token resolved per request from `provider()` (for
   * rotating/per-caller tokens).
   */
  static bearerTokenProvider(provider: BearerTokenProvider): Credentials {
    if (!provider) {
      throw new ConfigurationError("bearerTokenProvider is required");
    }
    return new Credentials(() => createBearerMiddleware(provider));
  }

  /** @internal Builds the authenticating OpenAPI middleware. */
  toMiddleware(): Middleware {
    return this.build();
  }
}
